processor/k8sattributesprocessor/internal/kube/client.go

```diff
@@ -374,12 +374,26 @@ func (c *WatchClient) extractPodAttributes(pod *api_v1.Pod) map[string]string {
 	return tags
 }
 
-func (c *WatchClient) extractPodContainersAttributes(pod *api_v1.Pod) map[string]*Container {
-	containers := map[string]*Container{}
-
-	if c.Rules.ContainerImageName || c.Rules.ContainerImageTag {
+func (c *WatchClient) extractPodContainersAttributes(pod *api_v1.Pod) PodContainers {
+	containers := PodContainers{
+		ByID:   map[string]*Container{},
+		ByName: map[string]*Container{},
+	}
+	if !needContainerAttributes(c.Rules) {
+		return containers
+	}
+	ifRule := func(rule bool, v string) string {
+		if rule {
+			return v
+		}
+		return ""
+	}
+	switch {
```

Comment:
Side note (cc @dmitryax): I would like to propose to remove all references to the rules from this package. That would then greatly simplify the complexity here. We can simply check the rules in the processor and avoid setting attributes if rules aren't set. This centralises and concentrates that logic in one places, making it much easier to reason about.

If performance was the concern and reason for the current architecture, then that can easily be addressed in another way.

If you like this idea, happy to propose & work on as a separate issue/PR.

Reply:
I'm always up for simplifying code! The only requirement is not to introduce any additional k8s API calls.  Please feel free to submit a separate PR

Reply:
Great. Will do. Thanks.